codex-rs/exec-server/src/posix/escalate_server.rs

```diff
@@ -0,0 +1,264 @@
+use std::collections::HashMap;
+use std::os::fd::AsRawFd;
+use std::path::Path;
+use std::path::PathBuf;
+use std::process::Stdio;
+use std::time::Duration;
+
+use anyhow::Context as _;
+
+use codex_core::exec::SandboxType;
+use codex_core::exec::process_exec_tool_call;
+use codex_core::get_platform_sandbox;
+use codex_core::protocol::SandboxPolicy;
+use tokio::process::Command;
+
+use crate::posix::escalate_protocol::BASH_EXEC_WRAPPER_ENV_VAR;
+use crate::posix::escalate_protocol::ESCALATE_SOCKET_ENV_VAR;
+use crate::posix::escalate_protocol::EscalateAction;
+use crate::posix::escalate_protocol::EscalateClientMessage;
+use crate::posix::escalate_protocol::EscalateServerMessage;
+use crate::posix::escalate_protocol::SuperExecMessage;
+use crate::posix::escalate_protocol::SuperExecResult;
+use crate::posix::socket::AsyncDatagramSocket;
+use crate::posix::socket::AsyncSocket;
+
+/// This is the policy which decides how to handle an exec() call.
+///
+/// `file` is the absolute, canonical path to the executable to run, i.e. the first arg to exec.
+/// `argv` is the argv, including the program name (`argv[0]`).
+/// `workdir` is the absolute, canonical path to the working directory in which to execute the
+/// command.
+pub(crate) type ExecPolicy = fn(file: &Path, argv: &[String], workdir: &Path) -> EscalateAction;
+
+pub(crate) struct EscalateServer {
+    bash_path: String,
+    policy: ExecPolicy,
+}
+
+impl EscalateServer {
+    pub fn new(bash_path: String, policy: ExecPolicy) -> Self {
+        Self { bash_path, policy }
+    }
+
+    pub async fn exec(
+        &self,
+        command: String,
+        env: HashMap<String, String>,
+        workdir: PathBuf,
+        timeout_ms: Option<u64>,
+    ) -> anyhow::Result<ExecResult> {
+        let (escalate_server, escalate_client) = AsyncDatagramSocket::pair()?;
+        let client_socket = escalate_client.into_inner();
+        client_socket.set_cloexec(false)?;
+
+        let escalate_task = tokio::spawn(escalate_task(escalate_server, self.policy));
+        let mut env = env.clone();
+        env.insert(
+            ESCALATE_SOCKET_ENV_VAR.to_string(),
+            client_socket.as_raw_fd().to_string(),
+        );
+        env.insert(
+            BASH_EXEC_WRAPPER_ENV_VAR.to_string(),
+            format!("{} escalate", std::env::current_exe()?.to_string_lossy()),
```

Comment:
As discussed, this is currently problematic if `std::env::current_exe()` contains a space.

Reply:
I would prefer we move to a world where `BASH_EXEC_WRAPPER_ENV_VAR` has to be just the executable and no args.